### PR TITLE
KAFKA-4049: Fix transient failure in RegexSourceIntegrationTest

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
ping @dguy @enothereska @mjsax @ijuma @ewencp for reviews.

File: clients/src/test/java/org/apache/kafka/test/TestUtils.java [hunk 1]
```diff
@@ -247,12 +247,12 @@ public static void waitForCondition(TestCondition testCondition, String conditio
...
 
-
-        while (!testCondition.conditionMet() && ((System.currentTimeMillis() - startTime) < maxWaitMs)) {
+        boolean conditionMet;
+        while (!(conditionMet = testCondition.conditionMet()) && ((System.currentTimeMillis() - startTime) < maxWaitMs)) {
             Thread.sleep(Math.min(maxWaitMs, 100L));
         }
 
-        if (!testCondition.conditionMet()) {
+        if (!conditionMet) {
```
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/test/TestUtils.java]: Is this change really required? It seems like pretty much any test that requires it is going to be fragile as it implies that the condition may only be met for < 100ms. Normally I would expect anything using `waitForCondition` to expect that condition to switch to true and stay in that state until the test returns from `waitForCondition` and takes further action such that the condition could change.

**Comment by ijuma** [comment]
Jenkins failed on  `org.apache.kafka.streams.integration.RegexSourceIntegrationTest.testRegexMatchesTopicsAWhenDeleted`

File: clients/src/test/java/org/apache/kafka/test/TestUtils.java (hunk 1 above)
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/test/TestUtils.java]: Agreed.

If we did hit this problem, then perhaps we should throw an exception if the condition changes between both checks with a message about fixing the condition to make the tests less fragile.

Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/test/TestUtils.java]: I think I'm convinced, will change it back.

**Comment by guozhangwang** [comment]
Discussed with @hachikuji about this issue, and it may be the result of KAFKA-3949, but it is a bit hard to confirm as I have not been able to re-produce locally. So I'll hold on this PR until that issue is fixed to see if it does resolve the transient failure.

**Comment by guozhangwang** [comment]
Waiting for consecutive jenkins builds on https://github.com/apache/kafka/pull/1762 now.

**Comment by guozhangwang** [comment]
#1762 seems not the root cause of this failure, will continue investigating in this PR.

**Comment by guozhangwang** [comment]
Triggered Jenkins build 5181 - 5185 for this commit.

**Comment by guozhangwang** [comment]
Apache Jenkins builds are not from expected commit, using  https://jenkins.confluent.io/job/kafka-branch-builder/ instead, with consecutive 74 - 78

**Comment by guozhangwang** [comment]
Consecutive tests in https://jenkins.confluent.io/job/kafka-branch-builder/ from 78 to 82. @ijuma mind taking a look again?

File: streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java [hunk 2]
```diff
@@ -209,34 +200,29 @@ public void testRegexMatchesTopicsAWhenDeleted() throws Exception {
...
 
         CLUSTER.deleteTopic("TEST-TOPIC-A");
 
-
         TestCondition oneTopicRemoved  = new TestCondition() {
             @Override
             public boolean conditionMet() {
-                List<String> assignedTopics = testStreamThread.assignedTopicPartitions.get(SECOND_UPDATE);
-                return assignedTopics != null && !assignedTopics.contains("TEST-TOPIC-A") && assignedTopics.contains("TEST-TOPIC-B");
+                System.out.println("deleted: " + testStreamThread.assignedTopicPartitions);
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java]: This was probably left by mistake.

**Comment by ijuma** [comment]
@guozhangwang it's worth noting that this test never failed in Confluent Jenkins (as far as I could see), only Apache Jenkins. Apache Jenkins runs the tests with fork=1 while Confluent uses a higher fork count. Usually this makes Apache Jenkins more stable, but it may be having the opposite effect in this case (or there's another reason why the test fails in Apache Jenkins). If you want to be confident that this change fixes the issue, we'd want to run it in Apache Jenkins.

File: streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java (hunk 2 above)
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java]: Yes, I'm aware of the changes and I intended to delete them before merging after your reviews. I should be more explicit on these two prints.

**Comment by ijuma** [comment]
This failure became quite rare at some point, last failure in trunk jdk7 build was a while back. Since this PR seems like an improvement anyway, LGTM (assuming the tests pass). If the test fails again, we can reopen the JIRA.

**Comment by guozhangwang** [comment]
Transient failure in Jenkins is not relevant; created https://issues.apache.org/jira/browse/KAFKA-4075 for it.

**Comment by guozhangwang** [comment]
Thanks @ijuma , merged to trunk.

**Pull request closed without merging**